34.md

```diff
@@ -0,0 +1,79 @@
+NIP-34
+======
+
+`git` stuff
+-----------
+
+`draft` `optional`
+
+This NIP defines all the ways code collaboration using and adjacent to [`git`](https://git-scm.com/) can be done using Nostr.
+
+## Repository announcements
+
+Git repositories are hosted in Git-enabled servers, but their existence can be announced using Nostr events, as well as their willingness to receive patches, bug reports and comments in general.
+
+```jsonc
+{
+  "kind": 30617,
+  "content": "",
+  "tags": [
+    ["d", "<repo-id>"],
+    ["name", "<human-readable project name>"],
+    ["description", "brief human-readable project description>"],
+    ["head", "<sha1-hex>"], // the current repository commit head, as hex
+    ["web", "<url for browsing>"], // a webpage url, if the git server being used provides such a thing
+    ["clone", "<url for git-cloning>"], // a url to be given to `git clone` so anyone can clone it
+    ["patches", "<relay-url>"], // this signals that patches are welcome and will be searched for in this relay
+    ["issues", "<relay-url>"] // this signals that bug reports, questions, comments and so on are welcome
+  ]
+}
+```
+
+The tags `web`, `clone`, `patches` and `issues` can be specified multiple times. Except `d`, all tags are optional.
+
+## Patches
+
+Patches can be sent by anyone to any repository. Patches to a specific repository SHOULD be sent to the relays specified in that repository's announcement event's `"patches"` tag. Patch events SHOULD include an `a` tag pointing to that repository's announcement address.
+
+```jsonc
+{
+  "kind": 1617,
+  "content": "<patch>", // contents of <git format-patch>
+  "tags": [
+    ["a", "30617:<base-repo-owner-pubkey>:<base-repo-id>"]
+  ]
+}
+```
+
+## Branch merge
+
+To be defined.
+
```

Comment:
I would remove these lines and add some `## Outlook` at the bottom to mention what else this nip could cover. Furthermore are you thinking of a "merge request" event here? As the different clone urls in the original repo might use different branch names, the target needs one specific url to reference the branch to merge into. As branches can be renamed, these events have no absolute cryptographic reference unless we add one via the branches heads.

MRs call for code reviews ...


````suggestion
## Merge Request

Notify repository owner of an improvement under development.

```jsonc
{
  "kind": 31617,
  "content": "<Description using markdown>",
  "tags": [
    ["d", "<something unique / UUID>"],
    ["a", "30617:<base-repo-owner-pubkey>:<base-repo-id>"],
    ["title", "<merge request title>"], // 
    ["state", "[wip|ready|final]"], // wip: share an idea, knowing it's not ready to merge, ready: author thinks it can be merged, final: author doesn't intend to update the source repo after feedback
    ["source", "<url for git-cloning>", "<branch name>", "<current branch head sha1-hex>"], // 
    ["target", "<url for git-cloning>", "<branch name>", "<current branch head sha1-hex>"], // 
  ]
}
```

## Merge Request Review

```jsonc
{
  "kind": 2617,
  "content": "<summary/comment allowing markdown>",
  "tags": [
    ["a", "30617:<base-repo-owner-pubkey>:<base-repo-id>"],
    ["a", "31617:<merge-request-owner-pubkey>:<merge-request-d-tag>"],
    ["type", "[ACK|NACK|...]"],
    ["anchor", "<source-commit-id>", "<target-commit-id>"],
    ["source-comment", "<file-with-path>:<line>", "<comment>"],
    ["target-comment", "<file-with-path>:<line>", "<comment>"],
  ]
}
```

Both `a` tags are required. The other tags are optional. `type` and `anchor` may only be used once.

````

Reply:
Yes, the branch merge thing would probably have to specify a branch name and a URL from where to fetch it. The person calling for a branch merge must have a git server somewhere. I don't know the details.

I was also thinking of another kind for inline comments of files, but all of that can be done much later I think.

Reply:
The MR state was just me being creative. An issue would usually not include the fix and when you want to send a fix, you can do so with a patch. The MR with a final state would be better done with a patch but on the other hand, maybe you want to propose a 5GB change, so a patch wouldn't work but if you host a fork, you can propose the 5GB change that way.

Reply:
The `state` field is probably not the best way to approach this. For example, closing an issue or a PR doesn't have to publish a whole new Issue Event, having just one field changed. I think it's better to have a separate kind for closing issues and PRs.